### DISCUSSION TITLE
Proper Database Migrations with Copilot Pipeline

### DISCUSSION THREAD

**Discussion by DrewVartanian** [aws/copilot-cli#3628] [Q&A]
We're looking to integrate migrations into our Copilot Pipeline deploy process, but having some trouble figuring it out.  Our migrations run through PHP Doctrine with Symfony, and they need to be run from a running container with our built codebase.  The solution that we're thinking about is having a service or job that is built alongside our web and worker images, which would run our migration when deployed.  We would then specify the web and worker containers to be dependent on the migration container (https://aws.github.io/copilot-cli/docs/manifest/pipeline/#stages-deployments-dependson).  The main issue with this solution is that the web and worker containers would be dependent on the _start_ of the migration job/service, whereas we need it to be dependent on the _completion_ and _success_ of the service.  Is there any way to have the depends_on rely on the success rather than the start of the migration?  Also, what type of job or service would you recommend using (jobs seem always run on a schedule, rather than an immediate one time execution, and services seem to be geared towards ongoing operations, rather than a finite task)?  Lastly, are we totally off base, and is there a more standard solution to running migrations during a Copilot Pipeline deploy?

**Comment by DrewVartanian** [comment]
Never got a suggestion here, but I figured it was worth sharing our solution in case anyone else comes across this.  Since we wanted a shared repo for our scheduled jobs (see [discussions/3598](https://github.com/aws/copilot-cli/discussions/3598)), we ended up building that image in the buildspec.yml
```
tag=$(sed 's/:/-/g' <<<"${CODEBUILD_BUILD_ID##*:}-${env}" | rev | cut -c 1-128 | rev);
JOB_IMAGE=$AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com/our-app/job;
docker build --file ./app/dev/dev-setup/containers/_base/Dockerfile --target job \
    --build-arg PROJECT_DIR=/var/www/our-app \
    --build-arg APP_ENV=$env \
    -t $JOB_IMAGE:latest \
    -t $JOB_IMAGE:$tag .;
if [ ! "$job_list" = null ]; then
    docker push $JOB_IMAGE --all-tags;
fi
```
We also needed to point our scheduled jobs' manifest.yml files to the image in AWS ECR, rather than having the build described there
```
image:
    location: xxxxx.dkr.ecr.xxxxx.amazonaws.com/${COPILOT_APPLICATION_NAME}/job:${tag}
```
Since we now have the job image built and available in our buildspec, we can add migrations to our post build phase in the buildspec file:
```
post_build:
  commands:
    - echo "Running Migrations"
    - >
      for env in $pl_envs; do
        tag=$(sed 's/:/-/g' <<<"${CODEBUILD_BUILD_ID##*:}-${env}" | rev | cut -c 1-128 | rev)
        JOB_IMAGE=$AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com/our-app/job;
        docker run \
          -u our_user \
          -e PROJECT_DIR=/var/www/our-app \
          -e APP_ENV=$env \
          $JOB_IMAGE:$tag php /var/www/our-app/app/console doctrine:migrations:migrate;
      done;
```
This isn't ideal, as we need to specify arguments in both the manifest and the buildspec files, and the environmental variables we have access to are different between the two files (eg: $AWS_ACCOUNT_ID only available in the buildspec file, and $COPILOT_APPLICATION_NAME only available in the buildspec.  It also breaks the pattern of having every container's build defined in the manifests, which was nice for a clean and clear infrastructure as code.  That said, as it solves both the issue of sharing one repo for all scheduled jobs and allows us to run migrations at the appropriate place in the build/deploy process, it's what we landed on.

Again, this is what we pieced together, what made sense given our setup, and is by no means a perfect solution.  That said, I felt it was worth sharing in case it's helpful to anyone else trying to figure this out.  Any other suggestions, thoughts, or critiques are welcome.

*Notes:
 - In order to allow buildspec.yml the ability to push to ECR, a policy must be added to the IAM BuildProjectRole(s).
 - The container running migrations from the buildspec.yml didn't have access to our RDS database by default.  Someone else tackled resolving that issue, but since it was not trivial, I'm including their notes below:
```
* VPC/Subnet/Security group setup:
  - Developer Tools -> CodeBuild -> Build projects -> [YOUR BUILD PROJECT]
  - Edit -> Environment
  - Expand "Additional configuration" near the bottom
  - Select your VPC
  - Add a "private" subnet
    - Steps for creating a "private" subnet:
      - VPC -> NAT gateways -> Create NAT gateway
      - Subnet: select existing "public" subnet (one with a route table pointing to an Internet Gateway), Connectivity type: public, Elastic IP allocation ID: click Allocate Elastic IP, click Create NAT gateway
      - VPC -> Subnets -> Create subnet
      - Select your VPC, click Create subnet
      - VPC -> Route tables -> Create route table
      - Select VPC, click Create route table
      - Edit the newly created route table routes
      - Add route, Destination: 0.0.0.0/0, Target: newly create NAT, click Save changes
      - VPC -> Subnets, click on newly created subnet
      - Click Route table tab, click Edit route table association
      - Select newly created route table, click Save
  - Add the copilot generated security group (same as the containers have)
  - Click Update environment
  ```

**Reply by efekarakus** [reply]
Thank you so much @DrewVartanian for the detailed response 😍 
Apologies 🙇 , we completely missed the post earlier but I am glad you found a workaround! I agree that there would be ideally a more native way of doing migrations via the pipeline. The post seems to be related to https://github.com/aws/copilot-cli/issues/3007

I'll try to answer some of the questions above here:

>  Is there any way to have the depends_on rely on the success rather than the start of the migration? 

Is this the section you're looking for: https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/#image-depends-on?
For example, you can specify:
```
image:
  build: ./Dockerfile
  depends_on:
     migration: success

sidecars:
  migration: $AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com/our-app/job
```

> Also, what type of job or service would you recommend using (jobs seem always run on a schedule, rather than an immediate one time execution, and services seem to be geared towards ongoing operations, rather than a finite task)?

In the past, we have seen folks use the `copilot task run` command in the [`test_commands` section](https://aws.github.io/copilot-cli/docs/manifest/pipeline/#stages-test-cmds) of the pipeline to run migrations or `copilot svc exec` to run a migration post deployment: https://github.com/aws/copilot-cli/issues/2992#issuecomment-964654705

These are not ideal but it's another way of mitigating the request.

**Reply by jetaggart** [reply]
@efekarakus Can you explain the sidecar example more? If I set up a job that has as a schedule of "none", will this boot that job and run it at least once?

**Reply by jetaggart** [reply]
I think I've figured out a decent way to work around this right now. If you define a job with that's unscheduled:

```yaml
name: migration
type: Scheduled Job

on:
  schedule: "none"
```

You can run this after you build all of your containers in the `buildspec.yaml`. Just before the line that's autogenerated:

```yaml
phases:
  install:
    ...
  build:
    commands:
      - echo "Run your tests"
      # - make test
  post_build:
    commands:
      # after all of the building and pushing of containers
      # INSERT THIS LINE
      - ./copilot-linux task run -n db-migrations --follow --execution-role EXECUTION_ROLE --task-role TAKS_ROLE --image MIGRATION_IMAGE:latest --env-vars ENV_VARS --secrets SECRETS --security-groups SECURITY_GROUPS --app APP --env $env
      - ls -lah ./infrastructure
```

This will run after all the images are built and just before they get deployed. I generated this command with this:

```
copilot task run --generate-cmd JOB_MANIFEST_NAME
```

You can probably do some bash wizardry to generate the line on the fly to pick up any changes (warning: definitely **not** valid sh): `copilot task run --generate-cmd JOB_MANIFEST_NAME | gsub --command="" --command="whatever" | ... other gsubs and formatting | exec`

**Reply by efekarakus** [reply]
That looks great! thank you for sharing it with everyone here ❤️
